pufferlib/ocean/trash_pickup/trash_pickup.h

```diff
@@ -0,0 +1,596 @@
+            env->trash_positions[trash_index][0] = -1;
+            env->trash_positions[trash_index][1] = -1;
+
+            env->agent_carrying[agent_idx] = AGENT_IS_CARRYING;
+            env->grid[INDEX(env, x, y)].type = EMPTY;
+            env->grid[INDEX(env, x, y)].index = -1;
+            env->grid[INDEX(env, new_x, new_y)].type = AGENT;
+            env->grid[INDEX(env, new_x, new_y)].index = agent_idx;
+            env->agent_positions[agent_idx][0] = new_x;
+            env->agent_positions[agent_idx][1] = new_y;
+            add_reward(env, env->positive_reward);
+        } 
+        else if (cell_state_type == TRASH_BIN) 
+        {
+            if (carrying == AGENT_IS_CARRYING)
+            {
+                // Deposit trash into bin
+                env->agent_carrying[agent_idx] = AGENT_IS_NOT_CARRYING;
+                add_reward(env, env->positive_reward);
+            }
+            else
+            {
+                bool try_pull_bin = false;
+                int new_bin_x = new_x + move_dir_x;
+                int new_bin_y = new_y + move_dir_y;
+                if (new_bin_x >= 0 && new_bin_x < env->grid_size && new_bin_y >= 0 && new_bin_y < env->grid_size)
+                {
+                    int new_bin_cell_state_type = env->grid[INDEX(env, new_bin_x, new_bin_y)].type;
+                    if (new_bin_cell_state_type == EMPTY)
+                    {
+                        int bin_index = env->grid[INDEX(env, new_x, new_y)].index;
+                        env->bin_positions[bin_index][0] = new_bin_x;
+                        env->bin_positions[bin_index][1] = new_bin_y;
+
+                        env->grid[INDEX(env, new_x, new_y)].type = EMPTY;
+                        env->grid[INDEX(env, new_x, new_y)].index = -1;
+                        env->grid[INDEX(env, new_bin_x, new_bin_y)].type = TRASH_BIN;
+                        env->grid[INDEX(env, new_bin_x, new_bin_y)].index = bin_index;
+                    }
+                    else{
+                        try_pull_bin = true;
+                    }
+                }
+                else
+                {
+                    try_pull_bin = true;
+                }
+
+                if (try_pull_bin)
+                {
+                    // pull the bin if on map border instead of push
+                    int new_agent_x = new_x - move_dir_x * 2;
+                    int new_agent_y = new_y - move_dir_y * 2;
+                    if (new_agent_x >= 0 && new_agent_x < env->grid_size && new_agent_y >= 0 && new_agent_y < env->grid_size)
+                    {
+                        // Verify there isn't another agent already occupying this slot
+                        // If its trash, lets just say the agent grabs it and puts it in the trash, env logic should handle this no problem.
+                        if (env->grid[INDEX(env, new_agent_x, new_agent_y)].type == EMPTY || env->grid[INDEX(env, new_agent_x, new_agent_y)].type == TRASH) 
```

Comment:
perfect example of unreadable line because no variable caching

Reply:
Fixed (to be pushed).